### PR TITLE
[exporter/clickhouse] merge scope name and version into SpanAttributes column in trace table

### LINKED ISSUE #17408: [exporter/clickhouse] Add instrumentation scope to span attribute
### Component(s)

exporter/clickhouse

### Is your feature request related to a problem? Please describe.

Currently, attributes such as `otel.library.name`, `otel.library.version` are not added to span attributes. Those attributes are parts of `InstrumentationScope`, which is provided by opentelemetry https://opentelemetry.io/docs/reference/specification/logs/data-model/#field-instrumentationscope

### Describe the solution you'd like

It would be nice to include `otel.library.name` and `otel.library.version` in span attributes

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Frapschen**
**Description:** merge scope name and version into resource attribute column in trace table

resolve: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17408

### CODE REVIEW

**Review by hanjm — APPROVED** [review]

File: exporter/clickhouseexporter/exporter_logs.go [hunk 1]
```diff
@@ -117,12 +117,14 @@ func (e *logsExporter) pushLogsData(ctx context.Context, ld plog.Logs) error {
...
-	m := make(map[string]string, attributes.Len())
-	attributes.Range(func(k string, v pcommon.Value) bool {
-		m[k] = v.AsString()
-		return true
-	})
+func attributesToMap(attributes ...pcommon.Map) map[string]string {
+	m := make(map[string]string)
+	for _, item := range attributes {
+		item.Range(func(k string, v pcommon.Value) bool {
+			m[k] = v.AsString()
```
Comment (atoulme) [inline exporter/clickhouseexporter/exporter_logs.go:124]: v.AsString() will return "" if the value is anything but a String. You can try `v.AsRaw()` then marshal the value to String.
```suggestion
			m[k] = v.AsRaw()
```

File: exporter/clickhouseexporter/exporter_traces.go [hunk 2]
```diff
@@ -103,7 +104,7 @@ func (e *tracesExporter) pushTraceData(ctx context.Context, td ptrace.Traces) er
 						r.Name(),
 						traceutil.SpanKindStr(r.Kind()),
 						serviceName,
-						resAttr,
+						attributesToMap(resAttr, scopeAttr),
```
Comment (atoulme) [inline exporter/clickhouseexporter/exporter_traces.go:107]: you could do this once per scope, right?

File: exporter/clickhouseexporter/exporter_logs.go (hunk 1 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_logs.go:124]: > will return "" if the value is anything but a String.

its `v.Str()`?

File: exporter/clickhouseexporter/exporter_traces.go (hunk 2 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_traces.go:107]: yes

**Comment by dmitryax** [comment]
Scope attributes were recently added but not sure they used at all. Still worth adding, but I think the issue is about scope name and version, not scope attributes.

**Comment by Frapschen** [comment]
> Scope attributes were recently added to the OTel spec, but not sure they used yet. Still worth adding, but I think the issue is about scope name and version, not scope attributes.

thank your for pointing out my misunderstanding,  I have recoded, please review  @dmitryax

File: exporter/clickhouseexporter/exporter_logs.go (hunk 1 above)
Comment (atoulme) [reply exporter/clickhouseexporter/exporter_logs.go]: I don't remember. Please make sure to test with at least with one value that is not a string and that you get the right string when inserting.

File: exporter/clickhouseexporter/exporter_traces.go [hunk 3]
```diff
@@ -38,6 +38,15 @@ type tracesExporter struct {
 	cfg    *Config
 }
 
+const (
+	otelScopeName    = "otel.scope.name"
+	otelScopeVersion = "otel.scope.version"
```
Comment (dmitryax) [inline exporter/clickhouseexporter/exporter_traces.go:43]: those are defined in collector/semconv as `AttributeOtelScopeName` and `AttributeOtelScopeVersion`. Please use them

File: exporter/clickhouseexporter/exporter_traces.go [hunk 4]
```diff
@@ -38,6 +38,15 @@ type tracesExporter struct {
...
+
+	// those are deprecated aliases, but we still report them for backward compatibility reasons
+	otelLibraryName    = "otel.library.name"
+	otelLibraryVersion = "otel.library.version"
```
Comment (dmitryax) [inline exporter/clickhouseexporter/exporter_traces.go:47]: Why is the backward compatibility needed for something gone before even the clickhouse exporter is introduced? I'd suggest not to put redundant fields

File: exporter/clickhouseexporter/exporter_traces.go [hunk 5]
```diff
@@ -90,6 +90,9 @@ func (e *tracesExporter) pushTraceData(ctx context.Context, td ptrace.Traces) er
 				serviceName = v.Str()
 			}
 			for j := 0; j < spans.ScopeSpans().Len(); j++ {
+				resAttr[conventions.AttributeOtelScopeName] = spans.ScopeSpans().At(j).Scope().Name()
+				resAttr[conventions.AttributeOtelScopeVersion] = spans.ScopeSpans().At(j).Scope().Version()
```
Comment (dmitryax) [inline exporter/clickhouseexporter/exporter_traces.go:94]: I just realized that we are violating OTel schema's representation that has been followed so far in the clickhouse table schema. Ideally, we need to add new columns for these fields. @Frapschen, will this be too big of a breaking change for such an addition? If that's the case, span attributes would be a better place. A scope cannot be mapped 1:1 to a resource. One resource can have multiple instrumentation scopes.

Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_traces.go]: I prefer add them to span attribute directly.

**Review by dmitryax — APPROVED** [review]

File: exporter/clickhouseexporter/exporter_traces.go [hunk 6]
```diff
@@ -94,6 +94,8 @@ func (e *tracesExporter) pushTraceData(ctx context.Context, td ptrace.Traces) er
 				for k := 0; k < rs.Len(); k++ {
 					r := rs.At(k)
 					spanAttr := attributesToMap(r.Attributes())
+					spanAttr[conventions.AttributeOtelScopeName] = spans.ScopeSpans().At(j).Scope().Name()
+					spanAttr[conventions.AttributeOtelScopeVersion] = spans.ScopeSpans().At(j).Scope().Version()
```
Comment (dmitryax) [inline exporter/clickhouseexporter/exporter_traces.go:98]: Should we check if name and version values are not empty before setting them?

**Review by dmitryax — COMMENTED** [review]
I believe we need to do the same for metrics and logs?

**Review by dmitryax — APPROVED** [review]
I think we should do the same for metrics and logs next. @Frapschen WDYT?

**Comment by Frapschen** [comment]
> I think we should do the same for metrics and logs next. @Frapschen WDYT?

yes, I think so.

**Pull request merged**